8342967: Lambda deduplication fails with non-metafactory BSMs and mismatched local variables names

biboudis opened the pull request:
After experimentation and under certain conditions, a few equivalent lambdas (up to variable renaming) fail to deduplicate.

`TreeHasher` and `TreeDiffer` are now aware that other bootstraps may appear in lambda bodies (e.g. `SwitchBootstraps.typeSwitch`) and that variable names do not matter (when they appear in method/lambda definitions). In the first case equivalence is checked based on `bsmKey` and not the `Dynamic{Var, Method}Symbol` itself.

The test was also adjusted since it was assuming BSM with certain structure only (that `.get(1)` was unprotected).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342967](https://bugs.openjdk.org/browse/JDK-8342967): Lambda deduplication fails with non-metafactory BSMs and mismatched local variables names (**Bug** - P4)


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21687/head:pull/21687` \
`$ git checkout pull/21687`

Update a local copy of the PR: \
`$ git checkout pull/21687` \
`$ git pull https://git.openjdk.org/jdk.git pull/21687/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21687`

View PR using the GUI difftool: \
`$ git pr show -t 21687`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21687.diff">https://git.openjdk.org/jdk/pull/21687.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21687#issuecomment-2435617388)
</details>

biboudis commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TreeDiffer.java:
```diff
@@ -197,13 +201,27 @@ public void visitIdent(JCIdent tree) {
+        Symbol symbol = tree.sym;
+        Symbol otherSymbol = that.sym;
+
+        result &= scan(symbol, otherSymbol);
```
Thanks! Done!

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TreeDiffer.java:
```diff
@@ -197,13 +201,28 @@ public void visitIdent(JCIdent tree) {
+        result = scan(symbol, otherSymbol);
+    }
+
+    private boolean scan(Symbol symbol, Symbol otherSymbol) {
```
Consider renaming this to `scanSymbol`, to make it clear it doesn't apply to trees.

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TreeDiffer.java:
```diff
@@ -667,10 +686,14 @@ public void visitVarDef(JCVariableDecl tree) {
                         && scan(tree.vartype, that.vartype)
                         && scan(tree.init, that.init);
+
+        if (!tree.sym.owner.type.hasTag(TypeTag.METHOD) && result) {
```
I would rewrite this (for improved readability) as:

```
JCVariableDecl that = (JCVariableDecl) parameter;
 result =
                scan(tree.mods, that.mods)
                        && scan(tree.nameexpr, that.nameexpr)
                        && scan(tree.vartype, that.vartype)
                        && scan(tree.init, that.init);

if (tree.sym.owner.type.hasTag(TypeTag.CLASS)) {
    // field names are important!
    result &= tree.name == that.name;
}

if (result) {
     equiv.put(tree.sym, that.sym);
}
```

mcimadamore replied:
And, now that I wrote that I realize there's another issue in the code (from before your changes). I don't think `equiv` is really meant to be for non-local variables. IMHO the logic should be:
1. if we're comparing two local variables, do a structural comparison (and ignore the names)
2. if we're comparing fields, just check the symbols are the same.
And... I'm not sure there's ever a case where (2) applies? E.g. do we really want to deduplicate lambdas containing similar local class definitions? @cushon what do you think? What was the original intention for this code?

lahodaj replied:
FWIW, I doubt we meaningfully deduplicate lambdas with local class definitions. I suspect we deduplicate them only if the declarations are not used, like:
```
         Runnable r1 = () -> {class C {}};
         Runnable r2 = () -> {class C {}};
```

But not when they are used, like:
```
         Runnable r1 = () -> {class C {} new C();};
         Runnable r2 = () -> {class C {} new C();};
```

because the use site (`new C();`) uses different `Symbol`s, and these `Symbol`s are not `equiv`-ed.

So, possibly, simply using `false` once we see a `JCClassDecl` might be a more direct, open, answer.

(Besides aliasing local variables, there are probably other cases where we don't detect semantically equivalent duplicates. Like `Object` and `java.lang.Object`. It is not clear to me if those are important in practice, though.)

mcimadamore replied:
@lahodaj  that's my feeling too - but if that's the case, I think all our dedup visitors should be more transparent that that's the case (which you seem to agree with).

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TreeHasher.java:
```diff
@@ -97,12 +101,21 @@ public void visitIdent(JCIdent tree) {
             }
         }
-        hash(sym);
+        if (sym instanceof PoolConstant.Dynamic dynamic) {
```
You could in principle add an `hashSymbol` method which does this dance - then you reuse it from both `visitIdent` and `visitSelect`

mcimadamore replied:
I'm not sure if `visitVarDef` is ok here. Again, we seem to assume that all variables encountered here are locals - and we use a number to hash them. This is fine for locals, but I think a field should just hash to the field symbol's hash? @cushon what do you think?

vicente-romero-oracle commented:
side: not sure if the dedup effort is worthy at the end in terms of performance / class file reduction

cushon commented:
> > side: not sure if the dedup effort is worthy at the end in terms of performance / class file reduction / maintenance
> 
> Good point - these deduplication tests are rather painful to maintain every time the compiler is changed (I had to fight with them as well when writing the revamped backend support for flexible constructor bodies). I share the general feeling of a slightly negative cost vs. benefit ratio -- but I don't have any backing evidence (maybe we should gather that up).

I don't have good data on what the remaining benefit is either, but for what it's worth I share the feeling that this feature may not be worth keeping around. The deduplication isn't performed if debug info is enabled to avoid incorrect line numbers in stack traces, so there's a large fraction of code that enables debug info and isn't benefiting from it.

vicente-romero-oracle commented:
> > side: not sure if the dedup effort is worthy at the end in terms of performance / class file reduction / maintenance
> 
> Good point - these deduplication tests are rather painful to maintain every time the compiler is changed (I had to fight with them as well when writing the revamped backend support for flexible constructor bodies). I share the general feeling of a slightly negative cost vs. benefit ratio -- but I don't have any backing evidence (maybe we should gather that up).

it seems like for some code bases it makes sense, which is I think why Liam proposed this issue, but this type of optimization seems to belong more to the VM than to javac

vicente-romero-oracle commented:
I was reading the discussion back then and it seems like the number of lambdas that could be deduplicated per file were < 1 on average, this is for their code base. We are not de-duplicating serializable lambdas and if we are generating debug info, then we are not deduplicating either

biboudis commented:
I addressed the review. One thing that I didn't yet change is `visitVarDef` after @mcimadamore's https://github.com/openjdk/jdk/pull/21687#discussion_r1816527035

@cushon what do you think?

biboudis commented:
In this [commit](https://github.com/openjdk/jdk/pull/21687/commits/afed36490a76d858b4909efb9fd5dc1487699cd2) I have added two groups of the same lambda (syntactically). There are
two checks in `DeduplicationTest.TreeDiffHashTaskListener.finished` as described in
the comments:

- // Scan the compilation for calls to a varargs method named 'group', whose arguments
   // are a group of lambdas that are equivalent to each other, but distinct from all
   // lambdas in the compilation unit outside of that group.
- // 1. Assert that all pairwise combinations of lambdas in the group are equal, and
   // hash to the same value.
- // 2. Assert that no lambdas in a group are equal to any lambdas outside that group,
   // or hash to the same value as lambda outside the group.

As a result, I added two groups (since there is no functionality to assert that
two lambdas in a group are different.) IIUC any lambda inside the group is
checked for non equality with each lambda outside the group.

        group((Function<Object, Integer>) x -> {class C {} new C(); return 42; });
        group((Function<Object, Integer>) x -> {class C {} new C(); return 42; });

I believe that now (since the same lambda is not equal to itself if it contains a class decl)
a reasonable if-test within the group, could be to test `if (rhs != lhs)`. Applying TreeDiff 
between the same lambda can yield an assertion error now. 

Let me know if I you believe that I misinterpreted the test in anyway.

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TreeHasher.java:
```diff
@@ -91,6 +92,12 @@ public void visitLiteral(JCLiteral tree) {
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        hash(tree.sym);
+        super.visitClassDef(tree);
```
I would remove the recursive call here - the hash of a class is just that of its symbol - no need to scan its members recursively.

mcimadamore commented on test/langtools/tools/javac/lambda/deduplication/Deduplication.java:
```diff
@@ -187,6 +187,9 @@ void f() {}
                     return y2;
                });
+
+        group((Function<Object, Integer>) x -> {class C {} new C(); return 42; });
```
It feels like the test logic could be maybe cleaner if `group` had two variants `groupEquals` and `groupNotEquals` - e.g. in this case you want to make sure the lambdas in the group do NOT dedup.

vicente-romero-oracle commented:
corpus jobs finished, mostly same results for both jobs one with and one without this patch. Of a total of ~23940 lambdas, none of them were deduplicated. This is a bit surprising tbh. It could be that most artifacts (all?) are generated with debug information. But still it seems like the benefit of this feature is marginal

biboudis commented:
Addressed the reviews with https://github.com/openjdk/jdk/pull/21687/commits/1c80a7c45e950cd49f98f22f1ec055be10e5f22e 
(which would be great if someone could have a final quick look).

I guess the only thing now is to decide (based on the experiments' results) whether we merge this and remove later, or close this and remove dedup now or remove later 😄

Sooner or later, this test will start to fail when lambda bodies start containing other kinds of bootstrap invocations.

lahodaj commented:
> corpus jobs finished, mostly same results for both jobs one with and one without this patch. Of a total of ~23940 lambdas, none of them were deduplicated. This is a bit surprising tbh. It could be that most artifacts (all?) are generated with debug information. But still it seems like the benefit of this feature is marginal

Maybe not so surprising. I believe the default for javac is `-g:lines`, which disables the de-duplication. So, it would actually require the builds to explicitly use `-g:none`, which is probably not very common.

biboudis commented:
/integrate
